sdks/java/extensions/sketching/src/main/java/org/apache/beam/sdk/extensions/sketching/MostFrequent.java

```diff
@@ -0,0 +1,747 @@
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.beam.sdk.extensions.sketching;
+
+import static com.google.common.base.Preconditions.checkArgument;
+
+import com.clearspring.analytics.stream.Counter;
+import com.clearspring.analytics.stream.StreamSummary;
+import com.google.auto.value.AutoValue;
+import java.io.Externalizable;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.ObjectInput;
+import java.io.ObjectInputStream;
+import java.io.ObjectOutput;
+import java.io.ObjectOutputStream;
+import java.io.OutputStream;
+import java.io.Serializable;
+import java.util.ArrayList;
+import java.util.Iterator;
+import java.util.List;
+import org.apache.beam.sdk.annotations.Experimental;
+import org.apache.beam.sdk.coders.Coder;
+import org.apache.beam.sdk.coders.CoderException;
+import org.apache.beam.sdk.coders.CustomCoder;
+import org.apache.beam.sdk.coders.KvCoder;
+import org.apache.beam.sdk.coders.VarLongCoder;
+import org.apache.beam.sdk.transforms.Combine;
+import org.apache.beam.sdk.transforms.Combine.CombineFn;
+import org.apache.beam.sdk.transforms.DoFn;
+import org.apache.beam.sdk.transforms.PTransform;
+import org.apache.beam.sdk.transforms.ParDo;
+import org.apache.beam.sdk.transforms.display.DisplayData;
+import org.apache.beam.sdk.util.CoderUtils;
+import org.apache.beam.sdk.values.KV;
+import org.apache.beam.sdk.values.PCollection;
+
+/**
+ * {@code PTransform}s for computing the most frequent elements of a stream.
+ *
+ * <p>This class uses the Stream Summary sketch which uses a HashMap to keep track of
+ * a certain number of elements considered as the most frequent elements in a stream.
+ *
+ * <h2>Brief overview of the sketch</h2>
+ *
+ * <p>The Stream Summary uses a doubly linked-list of buckets in order to order elements
+ * by frequency. Each bucket is associated with a list of Counters, each of them storing
+ * information about one of the tracked elements : value, count and maximum potential error.
+ * <br>
+ *   An element is guaranteed to be in the top K most frequent if its guaranteed number
+ *   of hits, i.e. the {@code count minus error}, is greater than the count of the element at
+ *   the position k+1.
+ * <br>
+ *   <b>WARNING</b>: Information about the error is lost during the parallelization,
```

Comment:
Does this mean that there are no guarantees whatsoever about the output of this algorithm, if merging of accumulators ever happens? I think that means that, until this issue is fixed in stream-lib, we need to not use Combine, and use a perhaps slower implementation but one that provides some guarantees.

Reply:
There's indeed no guarantee on the error.. But depending on the use case this should not be as awful as it looks...

For instance, the whole algorithm (and not just the implementation) is a problem for global windows because once a capacity is reached, it seems unlikely that an untracked elements goes all the way to the top and even in the case it would, there would be a high error.

However in the context of short timing windows, like hours or maybe days depending on the volume we can support, top elements if they appear soon enough in the process will be directly on top and the error will usually be 0 because they will never be discarded.

So yeah the output depends on the order of the element, but in our case I should rather say it depends on the timeliness more than the "exact" order

In all cases I already found some resources which present Algorithms adapted to parallélisation, so I will investigate on that soon.

Reply:
I looked up the algorithm and it says that it is guaranteed to find all items with count > n/k. Is this not the case in the current implementation? If so, is that because of merging or because of something else? If it's because of merging, let's get rid of merging - that will require not using a combine fn.

Reply:
Indeed it guaranteed to find all items with count > ØN, Ø being the support between 0 and 1. 
But this gives no guarantee about the output top k elements, which are of interest here. 
Thus the guarantee on the top k depends on the estimated error e, but due to the inconsistent merging this is not possible to compute.